### PR TITLE
[ci][config/03] store pr test results into s3 bucket

### PR DESCRIPTION
**Pull request by can-anyscale**
Add state machine aws bucket for pr branch. We need a separated bucket because pr has less permissions. We gather this additional data because why not ;). Seriously though, test results data gathered from PR will help us build another way of test selection algorithm.

Also store the data into s3 bucket.

Test:
- CI

### CODE REVIEW

**Comment by can-anyscale** [comment]
not working yet

**Review by aslonnie — COMMENTED** [review]
Have you already setup the bucket? I worry about the security aspects of this.

this is essentially giving the public free storage.

can we save the report in build job artifacts, and have a background job scanning and collecting these artifacts instead?

**Comment by can-anyscale** [comment]
i haven't yeah; how is the build job artifacts help with the public free storage issues?

**Comment by can-anyscale** [comment]
yea let's talk more tomorrow; whatever the trick the build artifacts are using, we can replicate that as well (obscuring bucket name, write-only-not-read, etc?)

**Comment by aslonnie** [comment]
> i haven't yeah; how is the build job artifacts help with the public free storage issues?

the build artifact has size limits, and also has retentions, and in our current way of doing things, it is also limited by the local disk size.

s3 writes are going to be unlimited..

**Comment by can-anyscale** [comment]
ok, let's do this first; in subsequent PR I'll improve the writing logic to s3 to go through the lambda function for extra verification

File: release/ray_release/util.py [hunk 1]
```diff
@@ -34,6 +39,35 @@ def __str__(self):
...
+    if bucket_v1:
+        return bucket_v1
+
+    pipeline_id = os.environ.get("BUILDKITE_PIPELINE_ID")
+    if pipeline_id in BRANCH_PIPELINES:
+        return get_global_config()["state_machine_branch_aws_bucket"]
+    if pipeline_id in PR_PIPELINES:
+        return get_global_config()["state_machine_pr_aws_bucket"]
+
+    assert False, f"Test state machine is not supported for pipeline {pipeline_id}"
```
Comment (aslonnie) [inline release/ray_release/util.py:68]: just assert `pipeline_id in PR_PIPELINES` on line 65?

`assert False` feels like a misuse on assertion.

File: release/ray_release/util.py [hunk 2]
```diff
@@ -34,6 +39,35 @@ def __str__(self):
...
+    # branch. This is because pr and branch pipeline have different permissions, and we
+    # want data on branch pipeline being protected.
+    bucket_v1 = get_global_config()["state_machine_aws_bucket"]
+    if bucket_v1:
+        return bucket_v1
+
+    pipeline_id = os.environ.get("BUILDKITE_PIPELINE_ID")
+    if pipeline_id in BRANCH_PIPELINES:
+        return get_global_config()["state_machine_branch_aws_bucket"]
+    if pipeline_id in PR_PIPELINES:
```
Comment (aslonnie) [inline release/ray_release/util.py:65]: is the pr bucket used or not? if not used yet, why return here?

**Review by aslonnie — APPROVED** [review]

File: release/ray_release/util.py (hunk 2 above)
Comment (can-anyscale) [reply release/ray_release/util.py:65]: yess, let move to the next pr where it is setup/used

**Pull request merged**